proxy/src/main/java/com/velocitypowered/proxy/protocol/StateRegistry.java

```diff
@@ -195,38 +200,40 @@
 
   public static class PacketRegistry {
 
-    private static final IntObjectMap<ImmutableIntArray> LINKED_PROTOCOL_VERSIONS = new IntObjectHashMap<>();
+    private static final EnumMap<ProtocolVersion, ImmutableList<ProtocolVersion>> LINKED_PROTOCOL_VERSIONS = new EnumMap(ProtocolVersion.class);
 
     static {
-      LINKED_PROTOCOL_VERSIONS.put(MINECRAFT_1_9,
-          ImmutableIntArray.of(MINECRAFT_1_9_1, MINECRAFT_1_9_2, MINECRAFT_1_9_4));
-      LINKED_PROTOCOL_VERSIONS.put(MINECRAFT_1_9_4,
-          ImmutableIntArray.of(MINECRAFT_1_10, MINECRAFT_1_11, MINECRAFT_1_11_1));
-      LINKED_PROTOCOL_VERSIONS.put(MINECRAFT_1_12, ImmutableIntArray.of(MINECRAFT_1_12_1));
-      LINKED_PROTOCOL_VERSIONS.put(MINECRAFT_1_12_1, ImmutableIntArray.of(MINECRAFT_1_12_2));
-      LINKED_PROTOCOL_VERSIONS
-          .put(MINECRAFT_1_13, ImmutableIntArray.of(MINECRAFT_1_13_1, MINECRAFT_1_13_2));
+      LINKED_PROTOCOL_VERSIONS.put(MINECRAFT_1_9, ImmutableList.of(MINECRAFT_1_9_1, MINECRAFT_1_9_2, MINECRAFT_1_9_4));
+      LINKED_PROTOCOL_VERSIONS.put(MINECRAFT_1_9_4, ImmutableList.of(MINECRAFT_1_10, MINECRAFT_1_11, MINECRAFT_1_11_1));
+      LINKED_PROTOCOL_VERSIONS.put(MINECRAFT_1_12, ImmutableList.of(MINECRAFT_1_12_1));
+      LINKED_PROTOCOL_VERSIONS.put(MINECRAFT_1_12_1, ImmutableList.of(MINECRAFT_1_12_2));
+      LINKED_PROTOCOL_VERSIONS.put(MINECRAFT_1_13, ImmutableList.of(MINECRAFT_1_13_1, MINECRAFT_1_13_2));
     }
 
     private final Direction direction;
-    private final IntObjectMap<ProtocolVersion> versions = new IntObjectHashMap<>(16);
+    private final EnumMap<ProtocolVersion, ProtocolRegistry> versions;
     private boolean fallback = true;
 
     PacketRegistry(Direction direction) {
       this.direction = direction;
-      ProtocolConstants.SUPPORTED_VERSIONS
-          .forEach(version -> versions.put(version, new ProtocolVersion(version)));
+
+      EnumMap<ProtocolVersion, ProtocolRegistry> mutableVersions = new EnumMap(ProtocolVersion.class);
```

Comment:
Missing diamond operator here: we can avoid a cast below.

Reply:
Doesn't avoid cast. Still complains about needing to cast. http://prntscr.com/lkdkg6

Reply:
Perhaps `Collections.<ProtocolVersion, ProtocolRegistry>unmodifiableMap()` would work?

Reply:
The fix is to change:
 `private final EnumMap<ProtocolVersion, ProtocolRegistry> versions;` 
to 
`private final Map<ProtocolVersion, ProtocolRegistry> versions;`

Are we okay with leaving the map type not explicitly defined?

Reply:
Yes, that's ok.